bpo-37007: Implement socket.if_nametoindex() and socket.if_indextoname() on Windows

ZackerySpytz opened the pull request:
<!-- issue-number: [bpo-37007](https://bugs.python.org/issue37007) -->
https://bugs.python.org/issue37007
<!-- /issue-number -->

ZackerySpytz committed:
Implement socket.if_nametoindex() and socket.if_indextoname() on Windows

zooba reviewed (approved):
Approved, but I asked on bpo whether it's worth doing if_nameindex as well.

ZackerySpytz committed:
Implement socket.if_nameindex() on Windows

testInterfaceNameIndex() in test_socket.py is now run on Windows
(as hasattr(socket, 'if_nameindex') is now True).
Two other tests that aren't directly related to if_nameindex() have
been disabled on Windows. These tests only use if_nameindex() for the
purpose of picking a random network interface.

zooba commented on Modules/socketmodule.c:
```diff
@@ -6588,28 +6590,54 @@ Set the default timeout in seconds (float) for new socket objects.\n\
+    PMIB_IF_TABLE2 tbl;
+    if (GetIfTable2Ex(MibIfTableRaw, &tbl) != NO_ERROR) {
+        Py_DECREF(list);
+        return PyErr_SetFromErrno(PyExc_OSError);
```
I think we want `SetFromWindowsError` here (or something like that). `SetFromErrno` will use the POSIX `errno` variable, while we want to use `GetLastError()`.

ZackerySpytz replied:
Apparently, `GetLastError()` can't be used with [GetIfTable2Ex()](https://docs.microsoft.com/en-us/windows/desktop/api/netioapi/nf-netioapi-getiftable2ex#return-value) or [ConvertInterfaceLuidToNameW()](https://docs.microsoft.com/en-us/windows/desktop/api/netioapi/nf-netioapi-convertinterfaceluidtonamew#return-value). The return value of those two functions must be used (I've verified this locally).

zooba commented on Modules/socketmodule.c:
```diff
@@ -6588,28 +6590,54 @@ Set the default timeout in seconds (float) for new socket objects.\n\
+                                        NDIS_IF_MAX_STRING_SIZE + 1)) {
+            Py_DECREF(list);
+            FreeMibTable(tbl);
+            return PyErr_SetFromErrno(PyExc_OSError);
```
Same here, but we'll need to capture `GetLastError()` in a variable before calling `FreeMibTable`.

ZackerySpytz replied:
See above.

zooba commented on Modules/socketmodule.c:
```diff
@@ -6588,28 +6590,54 @@ Set the default timeout in seconds (float) for new socket objects.\n\
+            return PyErr_SetFromErrno(PyExc_OSError);
+        }
+        PyObject *tuple = Py_BuildValue("IN", r.InterfaceIndex,
+                PyUnicode_FromWideChar(buf, -1));
```
`Py_BuildValue` can take `u` and a `wchar_t*` and will do the string conversion for you.

ZackerySpytz replied:
Done.

ZackerySpytz committed:
Fix the error handling in if_nameindex().

Use "u" with Py_BuildValue().

ZackerySpytz committed:
Merge remote-tracking branch 'upstream/master' into bpo-37007-socket-if_nametoindex-win

zooba commented on Modules/socketmodule.c:
```diff
@@ -6668,28 +6670,54 @@ Set the default timeout in seconds (float) for new socket objects.\n\
+    int ret;
+    if ((ret = GetIfTable2Ex(MibIfTableRaw, &tbl)) != NO_ERROR) {
+        Py_DECREF(list);
+        return PyErr_SetFromWindowsErr(ret);
```
```suggestion
        // ret is used instead of GetLastError()
        return PyErr_SetFromWindowsErr(ret);
```

I had to double check this, so probably worth calling it out. It's unusual for Windows APIs to return the error code.

ZackerySpytz replied:
Done.

zooba commented on Modules/socketmodule.c:
```diff
@@ -6668,28 +6670,54 @@ Set the default timeout in seconds (float) for new socket objects.\n\
+                                               Py_ARRAY_LENGTH(buf)))) {
+            Py_DECREF(list);
+            FreeMibTable(tbl);
+            return PyErr_SetFromWindowsErr(ret);
```
```suggestion
            // ret is used instead of GetLastError()
            return PyErr_SetFromWindowsErr(ret);
```

ZackerySpytz replied:
Done.

zooba reviewed (commented):
Just a suggested comment to avoid me having to look up the APIs again next time I read this :)

ZackerySpytz committed:
Apply suggestions from code review

Co-Authored-By: Steve Dower <steve.dower@microsoft.com>
